components/eam/src/physics/cam/modal_aer_opt.F90

```diff
@@ -1553,71 +1551,114 @@ end subroutine modal_size_parameters
+  real(r8), intent(in) :: table(ncoef,prefr,prefi)
+  real(r8), intent(in) :: ref_real(pcols), ref_img(pcols) !real and imganinary parts of refractive indices [unitless]
+  real(r8), intent(in) :: ref_real_tab(prefr), ref_img_tab(prefi) !!real and imganinary table refractive indices [unitless]
+
+  !intent-inouts/outs
+  integer,  intent(inout) :: itab(pcols), jtab(pcols)
+  real(r8), intent(inout) :: ttab(pcols), utab(pcols)
+  real(r8), intent(out)   :: coef(pcols,ncoef) !coefficient interpolated bilinearly
+
+  !local
+  integer  :: ip1, jp1, icoef, ic
+  real(r8) :: tu(pcols), tuc(pcols),tcu(pcols), tcuc(pcols)
+
+  if(itab(1) <= 0) then
+
+     !compute factors for the real part
+     call compute_factors(prefr, ncol, ref_real, ref_real_tab, & !in
+          itab, ttab) !out
+     !compute factors for the imaginary part
+     call compute_factors(prefi, ncol, ref_img, ref_img_tab, & !in
+          jtab, utab, .true.) !out
+
+  endif
+
+  !Interpolate coef
+  do ic = 1, ncol
+     tu(ic)   = ttab(ic)*utab(ic)
+     tuc(ic)  = ttab(ic)-tu(ic)
+     tcuc(ic) = 1._r8-tuc(ic)-utab(ic)
+     tcu(ic)  = utab(ic)-tu(ic)
+     jp1      = min(jtab(ic)+1,prefi)
+     ip1      = min(itab(ic)+1,prefr)
+     do icoef = 1, ncoef
+        coef(ic,icoef) = tcuc(ic)* table(icoef,itab(ic),jtab(ic)) + &
+             tuc(ic) * table(icoef,ip1,   jtab(ic)) + &
+             tu(ic)  * table(icoef,ip1,   jp1)    + &
+             tcu(ic) * table(icoef,itab(ic),jp1)
+     enddo
+  enddo
+end subroutine binterp
+
+
+subroutine compute_factors(prefri, ncol, ref_ind, ref_table, & !in
+     ix, tt, & !out
+     do_print) !in-optional
+
+  ! Compute factors for the real or imaginary parts
+  implicit none
+
+  integer, intent(in)  :: prefri, ncol
+  real(r8), intent(in) :: ref_table(:) !refractive index table [unitless]
+  real(r8), intent(in) :: ref_ind(:)   !refractive index       [unitless]
+
+  !FORTRAN refactor note: "do_print" is kept to maintain code consistenty with the
+  !original code. THis can be removed from the C++ ported code
+  logical, intent(in), optional :: do_print ! to print log msg or not
+
+  !intent-inouts/outs
+  integer,  intent(inout) :: ix(:)
+  real(r8), intent(inout) :: tt(:)
+
+  !local
+  integer  :: ii, ip1, ic
+  real(r8) :: dx
+
+  real(r8), parameter :: threshold = 1.e-20_r8
+
+  if(prefri > 1) then
+     do ic = 1, ncol
+        do ii = 1, prefri
+           if(ref_ind(ic) < ref_table(ii)) exit
+        enddo
+        ix(ic) = max(ii-1,1)
+        ip1    = min(ix(ic)+1,prefri)
+        dx     = (ref_table(ip1)-ref_table(ix(ic)))
+        if(abs(dx) > threshold) then
+           tt(ic) = (ref_ind(ic)-ref_table(ix(ic)))/dx
+           if (present(do_print) .and. do_print) then
+              if(tt(ic) < 0._r8 .or. tt(ic) > 1._r8) then
+                 write(iulog,*) 'tt,ref_ind,ix,ref_table,dx=',tt(ic),ref_ind(ic),ix(ic),ref_table(ix(ic)),dx
+              endif
+           endif
+        else
+           tt(ic) = 0._r8
```

Comment:
can we initiate tt(:) = 0._r8 at the beginning so we don't need the else condition here and below? 
It looks weird that all elements of tt (and likely ix) are overwritten here, why does it need to be an "inout" variable? Also they are not used in anywhere else in modal_aero_sw and modal_aero_lw. Can it be put as local variable in binterp?

Reply:
That's a good observation. I have removed both the `else` statements and simplified the code. I have also make `tt` and `ix` intent-out (as you suggested).
Both `tt` and `ix` (real and imaginary counterparts) are computed in the first call of binterp and used in the subsequent calls. We can make them local if we just had one binterp call (such as in long wave). For shortwave, we make three calls, so we need them for the second and the third call as input.

Another option would be to declare them with `save` or declare them globally at the module level. Both these solutions can cause issues with threading. So sending them via arguments is the best solution and it also makes it clear what the dependencies are by just looking at the argument list.